pkg/kubelet/kubelet_pods.go

```diff
@@ -509,30 +509,34 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string, enableServiceLinks bool) (map[
 		// Kubelets without masters (e.g. plain GCE ContainerVM) don't set env vars.
 		return m, nil
 	}
-	services, err := kl.serviceLister.List(labels.Everything())
-	if err != nil {
-		return m, fmt.Errorf("failed to list services when setting up env vars")
-	}
 
-	// project the services in namespace ns onto the master services
-	for i := range services {
-		service := services[i]
-		// ignore services where ClusterIP is "None" or empty
-		if !v1helper.IsServiceIPSet(service) {
-			continue
+	// add environment variables for other services in the same namespace,
+	// if enableServiceLinks is true.
+	if enableServiceLinks {
+		services, err := kl.serviceLister.Services(ns).List(labels.Everything())
+		if err != nil {
+			return m, fmt.Errorf("failed to list services when setting up env vars")
 		}
-		serviceName := service.Name
 
-		// We always want to add environment variabled for master services
-		// from the master service namespace, even if enableServiceLinks is false.
-		// We also add environment variables for other services in the same
-		// namespace, if enableServiceLinks is true.
-		if service.Namespace == kl.masterServiceNamespace && masterServices.Has(serviceName) {
-			if _, exists := serviceMap[serviceName]; !exists {
-				serviceMap[serviceName] = service
+		for _, service := range services {
```

Comment:
do this to avoid capturing the range variable in the return value (https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable)

```suggestion
		for i := range services {
			service := services[i]
```

Reply:
@liggitt there is no problem because the output of `serviceLister` is a array of pointers of v1.Service, the related problem above is when you use a pointer of array of object inside a loop.